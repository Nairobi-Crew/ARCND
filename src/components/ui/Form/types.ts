import { FC, FormHTMLAttributes } from 'react';

export type TOnSubmit = () => void;

export type OwnFormProps = {
  onSubmit?: TOnSubmit
  classList?: string[]
  caption?: string
  visible?: boolean
} & FormHTMLAttributes<HTMLFormElement>;

export type FormProps = FC<OwnFormProps>;
