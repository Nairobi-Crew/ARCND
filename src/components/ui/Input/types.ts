import { FC, InputHTMLAttributes } from 'react';

export type TOnValueChange = (value: string) => void;

export type OwnInputProps = {
  id: string
  label?: string
  type?: string
  labelClasses?: string[]
  inputClasses?: string[]
  onValueChanged?: TOnValueChange
} & InputHTMLAttributes<HTMLInputElement>;

export type InputProps = FC<OwnInputProps>;
