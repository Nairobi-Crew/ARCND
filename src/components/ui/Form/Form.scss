@import "../../../common/vars";

.form {

  &__container {
    display: flex;
    flex-direction: row;
    height: 100vh;
    align-items: center;
  }

  &__empty_column {
    flex: 1;
  }

  &__form_column {
    .caption {
      font-size: $f-xl;
      font-family: $ff;
      text-align: center;
      padding: 15px 30px;
      box-shadow: $shadow-small;
      text-transform: uppercase;
      margin: 0 0 30px;
    }

    align-items: center;
    align-content: center;
    background: $form-background-color;
    border-radius: $form-border-radius;
    box-shadow: $shadow;
    flex: 1;
    padding: 40px;

    @media only screen and (max-width: 768px) {
      flex: 5;
    }
  }
}
