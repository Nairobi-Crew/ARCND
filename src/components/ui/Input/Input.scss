@import "src/common/utils";

.input {
  font-weight: 400;
  height: 56px;
  width: 100%;
  background-color: inherit;
  @include font-color;
  margin: 0 0 40px;

  textarea {
    padding-top: 12px;
  }

  &__label {
    position: relative;
    display: block;
    height: 100%;
    width: 100%;
    font-family: $ff;
    @include font-color;
    background-color: inherit;
  }

  &__input {
    resize: none;
    height: 100%;
    width: 100%;
    outline: none;
    border: none;
    border-bottom: 2px solid #49005d;
    background-color: inherit;
    font-family: $ff;
    @include font-color;
  }

  &__input:focus {
    border-color: $primary-btn;
  }

  &__input:invalid:not(:placeholder-shown),
  &__input_error {
    border-color: $error;
  }

  &__input:valid:not(:placeholder-shown):not(:focus):not(&__input_error) {
    border-color: $secondary-btn;
  }

  &__input::placeholder {
    opacity: 0;
  }

  &__label-text {
    position: absolute;
    font-size: $f-m;
    font-family: $ff;
    top: 50%;
    left: 0;
    padding: 2px 8px 2px 2px;
    background-color: inherit;
    transform: translateY(-50%);
    transition: all 0.3s;
  }

  &__label-text > sup {
    position: absolute;
    bottom: 5px;
    right: 2px;
    color: $error;
  }

  &__input:not(:placeholder-shown) + &__label-text,
  &__input:focus + &__label-text {
    top: 0;
    font-size: 9px;
  }

  &__error-message {
    font-size: $f-m;
    color: $error;
  }
}
