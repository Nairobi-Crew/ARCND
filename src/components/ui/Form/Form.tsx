import React from 'react';
import { FormProps } from 'UI/Form/types';
import './Form.scss';
import Header from 'UI/Header/Header';

const Form: FormProps = ({ children, onSubmit, caption }) => {
  const onSubmitHandler = (evt) => {
    evt.preventDefault();
    if (onSubmit) {
      onSubmit();
    }
  };
  return (
    <div className="form__container">
      <div className="form__empty_column">
        <Header />
      </div>
      <div className="form__form_column">
        <h1 className="form__form_column caption">{caption}</h1>
        <form onSubmit={onSubmitHandler}>
          {children}
        </form>
      </div>
      <div className="form__empty_column" />
    </div>
  );
};

export default Form;
