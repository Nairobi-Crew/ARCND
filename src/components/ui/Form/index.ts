import { memo } from 'react';
import { OwnFormProps } from 'UI/Form/types';
import Form from 'UI/Form/Form';

export default memo<OwnFormProps>(Form);
