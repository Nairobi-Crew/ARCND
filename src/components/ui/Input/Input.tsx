import React, { useEffect, useState } from 'react';
import { InputProps } from 'UI/Input/types';
import './Input.scss';

const Input: InputProps = ({
  label, value, onValueChanged, errorMessage, ...restProps
}) => {
  const [inputValue, setInputValue] = useState<string>(value as string || '');
  useEffect(() => {
    setInputValue(value as string);
  }, [value]);
  const onInputHandler = (e: React.FormEvent<HTMLInputElement>): void => {
    if (e.target) {
      const v = (e.target as HTMLInputElement).value;
      setInputValue(v);
      if (onValueChanged) {
        onValueChanged(v);
      }
    }
  };

  return (
    <div className="input">
      <label className="input__label">
        <input
          className="input__input"
          value={inputValue}
          placeholder={label}
          onInput={onInputHandler}
          {...restProps}
        />
        <span className="input__label-text">{label}</span>
      </label>
      <span className="input__error-message">{errorMessage}</span>
    </div>
  );
};

export default Input;
