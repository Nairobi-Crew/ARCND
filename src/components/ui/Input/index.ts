import Input from 'UI/Input/Input';
import { memo } from 'react';
import { OwnInputProps } from 'UI/Input/types';

export default memo<OwnInputProps>(Input);
