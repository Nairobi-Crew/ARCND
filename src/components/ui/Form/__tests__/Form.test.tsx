import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import React from 'react';
import Form from 'UI/Form/Form';
import Input from 'UI/Input/Input';
import { BrowserRouter as Router } from 'react-router-dom';

describe('Render Form', () => {
  const Component = () => (
    <Router>
      <Form>
        <Input value="Test value" label="Test" />
      </Form>
    </Router>
  );

  it('Render form correct', () => {
    const inputComponent = renderer.create(<Component />).toJSON();
    expect(inputComponent).toMatchSnapshot();
  });

  it('test className', () => {
    const wrapper = mount(<Component />);
    expect(wrapper.find('.form__container')).toHaveLength(1);
    expect(wrapper.find('.form__form_column')).toHaveLength(2);
  });
});
