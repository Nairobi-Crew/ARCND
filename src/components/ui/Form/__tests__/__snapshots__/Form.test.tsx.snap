// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Form Render form correct 1`] = `
<div
  className="form__container  form__container_max_height"
>
  <div
    className="form__empty_column form__empty_column_max_height"
  >
    <header
      className="header"
    >
      <a
        className="header__title"
        href="/"
        onClick={[Function]}
      >
        <h1>
          ARCND
        </h1>
      </a>
    </header>
  </div>
  <div
    className="form__form_column"
  >
    <h1
      className="form__form_column caption"
    />
    <form
      onSubmit={[Function]}
    >
      <div
        className="input"
      >
        <label
          className="input__label"
        >
          <input
            className="input__input"
            onInput={[Function]}
            placeholder="Test"
            value="Test value"
          />
          <span
            className="input__label-text"
          >
            Test
          </span>
        </label>
        <span
          className="input__error-message"
        />
      </div>
    </form>
  </div>
  <div
    className="form__empty_column"
  />
</div>
`;
